File: packages/react-native-ekyc/src/liveness/challenges.ts
import type { ChallengeName, FaceSignal } from '../types'
import { Challenge, type ChallengeMemo, type ChallengeMetric } from './Challenge'

/**
 * Which way ML Kit's `yawAngle` grows when the user turns to *their own* left.
 *
 * Calibrated on a real Android phone (Android 35, front camera): ML Kit
 * reports **positive** yaw for a turn to the user's left, so `TurnLeft` must
 * accept positive values — hence -1 under the convention below (`yaw * sign
 * <= -minYaw`). Left as a knob because ML Kit's sign has flipped between
 * releases; re-check with `<EKYCCamera debug />` if instructions ever point
 * the wrong way. The web preview's MediaPipe adapter mirrors its yaw so the
 * same value works there.
 *
 * The server does not depend on this at all: it verifies that the two turns
 * went in *opposite* directions without naming either one (see the design
 * spec, "convention-free pose verification"). So getting it wrong costs you a
 * confusing instruction, not a security hole.
 */
export const DEFAULT_YAW_SIGN: 1 | -1 = -1

/**
 * Defaults, and where they come from.
 *
 * None of these is a magic number typed once and forgotten: each mirrors a
 * server rule (documented in `server/app/config.py`) plus a small margin, so
 * the phone confirms a pose only once the server would also accept it. When a
 * session comes from the server, `EKYCCamera` overrides them from the
 * `policy` the server issued (`tuningFromPolicy`), so the two never drift
 * apart. Every value is a *change from the person's own neutral frame*.
 */
export const CHALLENGE_DEFAULTS = {
  /** |yaw| and |pitch| the centre step accepts as "facing the camera". */
  centerMaxYaw: 12,
  centerMaxPitch: 12,
  /** Server: `turn_yaw_min_deg` 22 from neutral. Margin +3 so the mid-hold snapshot clears it. */
  turnMinYawDelta: 25,
  /** Nod, phase 1: pitch excursion from neutral in either direction. Local-only flow (the server never issues `nod`). */
  nodMinPitchDelta: 12,
  /** Nod, phase 2: excursion to the *opposite* side, as a fraction of `nodMinPitchDelta` — up 12° then down ≥ 7.2° = a real nod, ≥ 19° of travel. */
  nodReturnFraction: 0.6,
  /** Mouth: contour gap ratio *rise* over the neutral frame. Server checks jawOpen ≥ 0.35 & Δ ≥ 0.20; the contour ratio runs ~0 shut → 0.3+ open. */
  mouthOpenMinDelta: 0.18,
  /** Fallback when no baseline exists (a host that skips `center`). */
  mouthOpenMinAbsolute: 0.3,
  /** ML Kit mean eye-open probability at or below this = closed. Open reads 0.9+. */
  eyesClosedMaxOpen: 0.5,
  /** Blink, phase 2: eyes must read open again — a shut photo never does. */
  eyesReopenMinOpen: 0.7,
  /** Mouth, phase 2: after the held open frame the gap must fall back to within this fraction of the required rise — a gaping mask/photo never closes. */
  mouthCloseFraction: 0.5,
  smileMin: 0.7,
  /** Smile, phase 2: probability must drop back below this fraction of `smileMin`. */
  smileRelaxFraction: 0.5,
} as const

export type CenterOptions = {
  /** Max |yaw| and |pitch| in degrees that still counts as facing the camera. */
  maxYaw?: number
  maxPitch?: number
}

export type TurnOptions = {
  /** How far the head must rotate from the neutral frame, in degrees. */
  minYaw?: number
  yawSign?: 1 | -1
}

export type CloseEyesOptions = {
  /** Mean open-probability of the two eyes must be at or below this. */
  maxEyeOpen?: number
  /** ...and rise back to at least this afterwards (phase 2). */
  reopenMinOpen?: number
}

export type SmileOptions = {
  minSmile?: number
}

export type NodOptions = {
  /** Pitch excursion from the neutral frame (degrees), either direction, for phase 1. */
  minPitch?: number
  /** Opposite-side excursion for phase 2, as a fraction of `minPitch`. */
  returnFraction?: number
}

export type OpenMouthOptions = {
  /** Rise of `FaceSignal.mouthOpen` over the neutral frame. */
  minDelta?: number
  /** Absolute floor used only when there is no baseline. */
  minMouthOpen?: number
  /** Phase 2: the gap must fall back to baseline + this fraction of the rise. */
  closeFraction?: number
}

/**
 * Face the camera straight on.
 *
 * Always the first step: its captured frame is the `neutral` evidence the
 * server measures every other frame against, and the baseline every later
 * challenge on the phone is judged relative to.
 */
export class CenterChallenge extends Challenge {
  readonly name: ChallengeName = 'center'
  override readonly requiresRecenter = false

  constructor(private readonly options: CenterOptions = {}) {
    super()
  }

  isSatisfied(signal: FaceSignal): boolean {
    const { maxYaw = CHALLENGE_DEFAULTS.centerMaxYaw, maxPitch = CHALLENGE_DEFAULTS.centerMaxPitch } = this.options
    return Math.abs(signal.yaw) <= maxYaw && Math.abs(signal.pitch) <= maxPitch
  }

  override metric(signal: FaceSignal): ChallengeMetric {
    const { maxYaw = CHALLENGE_DEFAULTS.centerMaxYaw } = this.options
    return { value: Math.max(Math.abs(signal.yaw), Math.abs(signal.pitch)), needed: maxYaw, direction: 'below' }
  }
}

/**
 * Blink (or close the eyes).
 *
 * An event, not a hold: one frame with the eyes shut completes the step, so a
 * natural ~150 ms blink is enough. That is only sound because stills are
 * snapshots of the preview surface — grabbed the instant the closed frame is
 * seen, with no shutter lag — and because the *server* re-checks eye
 * closure on that frame against the neutral one. The device merely picks the
 * moment; it proves nothing.
 */
export class CloseEyesChallenge extends Challenge {
  readonly name: ChallengeName = 'closeEyes'
  readonly holdMs = 0
  /** closed → open again. The evidence frame is the closed one. */
  override readonly phaseCount = 2
  override readonly capturePhase = 0

  constructor(private readonly options: CloseEyesOptions = {}) {
    super()
  }

  isSatisfied(signal: FaceSignal, _baseline: FaceSignal | null = null, phase = 0): boolean {
    // ML Kit's open-probability lags a fast blink and the two eyes rarely
    // bottom out on the same frame, so require the *average* below the
    // threshold rather than both eyes independently.
    const { maxEyeOpen = CHALLENGE_DEFAULTS.eyesClosedMaxOpen, reopenMinOpen = CHALLENGE_DEFAULTS.eyesReopenMinOpen } = this.options
    const open = (signal.leftEye + signal.rightEye) / 2
    return phase === 0 ? open <= maxEyeOpen : open >= reopenMinOpen
  }

  override metric(signal: FaceSignal, _baseline: FaceSignal | null = null, phase = 0): ChallengeMetric {
    const { maxEyeOpen = CHALLENGE_DEFAULTS.eyesClosedMaxOpen, reopenMinOpen = CHALLENGE_DEFAULTS.eyesReopenMinOpen } = this.options
    const open = (signal.leftEye + signal.rightEye) / 2
    return phase === 0 ? { value: open, needed: maxEyeOpen, direction: 'below' } : { value: open, needed: reopenMinOpen, direction: 'above' }
  }
}

/** Yaw change from the baseline, positive towards the requested side. */
function turnDelta(signal: FaceSignal, baseline: FaceSignal | null, sign: 1 | -1, towards: 'left' | 'right'): number {
  const delta = signal.yaw - (baseline?.yaw ?? 0)
  return towards === 'left' ? -delta * sign : delta * sign
}

export class TurnLeftChallenge extends Challenge {
  readonly name: ChallengeName = 'turnLeft'

  constructor(private readonly options: TurnOptions = {}) {
    super()
  }

  isSatisfied(signal: FaceSignal, baseline: FaceSignal | null = null): boolean {
    const { minYaw = CHALLENGE_DEFAULTS.turnMinYawDelta, yawSign = DEFAULT_YAW_SIGN } = this.options
    return turnDelta(signal, baseline, yawSign, 'left') >= minYaw
  }

  override metric(signal: FaceSignal, baseline: FaceSignal | null = null): ChallengeMetric {
    const { minYaw = CHALLENGE_DEFAULTS.turnMinYawDelta, yawSign = DEFAULT_YAW_SIGN } = this.options
    return { value: turnDelta(signal, baseline, yawSign, 'left'), needed: minYaw, direction: 'above' }
  }
}

export class TurnRightChallenge extends Challenge {
  readonly name: ChallengeName = 'turnRight'

  constructor(private readonly options: TurnOptions = {}) {
    super()
  }

  isSatisfied(signal: FaceSignal, baseline: FaceSignal | null = null): boolean {
    const { minYaw = CHALLENGE_DEFAULTS.turnMinYawDelta, yawSign = DEFAULT_YAW_SIGN } = this.options
    return turnDelta(signal, baseline, yawSign, 'right') >= minYaw
  }

  override metric(signal: FaceSignal, baseline: FaceSignal | null = null): ChallengeMetric {
    const { minYaw = CHALLENGE_DEFAULTS.turnMinYawDelta, yawSign = DEFAULT_YAW_SIGN } = this.options
    return { value: turnDelta(signal, baseline, yawSign, 'right'), needed: minYaw, direction: 'above' }
  }
}

export class SmileChallenge extends Challenge {
  readonly name: ChallengeName = 'smile'
  /** smile (held, snapshotted) → relax again: a smiling photo never relaxes. */
  override readonly phaseCount = 2
  override readonly capturePhase = 0

  constructor(private readonly options: SmileOptions = {}) {
    super()
  }

  isSatisfied(signal: FaceSignal, _baseline: FaceSignal | null = null, phase = 0): boolean {
    const { minSmile = CHALLENGE_DEFAULTS.smileMin } = this.options
    return phase === 0 ? signal.smile >= minSmile : signal.smile <= minSmile * CHALLENGE_DEFAULTS.smileRelaxFraction
  }

  override metric(signal: FaceSignal, _baseline: FaceSignal | null = null, phase = 0): ChallengeMetric {
    const { minSmile = CHALLENGE_DEFAULTS.smileMin } = this.options
    return phase === 0
      ? { value: signal.smile, needed: minSmile, direction: 'above' }
      : { value: signal.smile, needed: minSmile * CHALLENGE_DEFAULTS.smileRelaxFraction, direction: 'below' }
  }
}

/**
 * Open the mouth — judged as a *rise* over the person's own resting mouth.
 *
 * The rigid-mask counter-measure: a 3-D-printed, resin or latex mask cannot
 * open its jaw, and a flexible silicone mask opens far less than the wearer's.
 * The phone only picks the moment; the server verifies `jawOpen` against the
 * neutral frame from MediaPipe blendshapes.
 */
export class OpenMouthChallenge extends Challenge {
  readonly name: ChallengeName = 'openMouth'
  /** open (held, snapshotted) → closed again. A gaping photo or mask never closes. */
  override readonly phaseCount = 2
  override readonly capturePhase = 0

  constructor(private readonly options: OpenMouthOptions = {}) {
    super()
  }

  private needed(baseline: FaceSignal | null): number {
    const { minDelta = CHALLENGE_DEFAULTS.mouthOpenMinDelta, minMouthOpen = CHALLENGE_DEFAULTS.mouthOpenMinAbsolute } = this.options
    return baseline ? baseline.mouthOpen + minDelta : minMouthOpen
  }

  private closedBelow(baseline: FaceSignal | null): number {
    const { minDelta = CHALLENGE_DEFAULTS.mouthOpenMinDelta, closeFraction = CHALLENGE_DEFAULTS.mouthCloseFraction } = this.options
    const rest = baseline ? baseline.mouthOpen : 0
    return rest + minDelta * closeFraction
  }

  isSatisfied(signal: FaceSignal, baseline: FaceSignal | null = null, phase = 0): boolean {
    return phase === 0 ? signal.mouthOpen >= this.needed(baseline) : signal.mouthOpen <= this.closedBelow(baseline)
  }

  override metric(signal: FaceSignal, baseline: FaceSignal | null = null, phase = 0): ChallengeMetric {
    return phase === 0
      ? { value: signal.mouthOpen, needed: this.needed(baseline), direction: 'above' }
      : { value: signal.mouthOpen, needed: this.closedBelow(baseline), direction: 'below' }
  }
}

/**
 * Nod: pitch the head clearly away from the resting pitch, either direction.
 *
 * Direction-free on purpose, like the turns: ML Kit's pitch sign differs
 * between devices, and what matters for liveness is that the head *moved*
 * on a second axis. Used by the local-only flow; the server never issues it.
 */
export class NodChallenge extends Challenge {
  readonly name: ChallengeName = 'nod'
  /**
   * Phase 1: pitch away from neutral by `minPitch` in either direction (the
   * memo remembers which). Phase 2: pitch to the *opposite* side by
   * `returnFraction · minPitch`. Up-then-down or down-then-up both count —
   * ML Kit's pitch sign differs between devices and what matters is the
   * movement through the resting pose. A held tilt, or a photo held at an
   * angle, never completes phase 2. The evidence frame is the phase-2 frame.
   */
  override readonly phaseCount = 2
  readonly holdMs = 0

  constructor(private readonly options: NodOptions = {}) {
    super()
  }

  isSatisfied(signal: FaceSignal, baseline: FaceSignal | null = null, phase = 0, memo: ChallengeMemo = {}): boolean {
    const { minPitch = CHALLENGE_DEFAULTS.nodMinPitchDelta, returnFraction = CHALLENGE_DEFAULTS.nodReturnFraction } = this.options
    const delta = signal.pitch - (baseline?.pitch ?? 0)
    if (phase === 0) {
      if (Math.abs(delta) < minPitch) return false
      memo.nodSign = Math.sign(delta)
      return true
    }
    const sign = memo.nodSign ?? 0
    if (sign === 0) return false
    return -sign * delta >= minPitch * returnFraction
  }

  override metric(signal: FaceSignal, baseline: FaceSignal | null = null, phase = 0, memo: ChallengeMemo = {}): ChallengeMetric {
    const { minPitch = CHALLENGE_DEFAULTS.nodMinPitchDelta, returnFraction = CHALLENGE_DEFAULTS.nodReturnFraction } = this.options
    const delta = signal.pitch - (baseline?.pitch ?? 0)
    if (phase === 0) return { value: Math.abs(delta), needed: minPitch, direction: 'above' }
    const sign = memo.nodSign ?? 0
    return { value: -sign * delta, needed: minPitch * returnFraction, direction: 'above' }
  }
}

export type ChallengeTuning = {
  center?: CenterOptions
  closeEyes?: CloseEyesOptions
  turn?: TurnOptions
  smile?: SmileOptions
  openMouth?: OpenMouthOptions
  nod?: NodOptions
}

/**
 * The thresholds a server session carries (`SessionPolicy`), turned into
 * challenge tuning: server rule + a margin, so what the phone confirms is what
 * the server will accept. Anything the policy does not carry keeps the local
 * default; explicit `tuning` from the host app wins over both.
 */
export function tuningFromPolicy(
  policy: { turnYawMinDeg?: number; neutralYawMaxDeg?: number } | undefined,
  tuning: ChallengeTuning = {},
): ChallengeTuning {
  if (!policy) return tuning
  const out: ChallengeTuning = { ...tuning }
  if (policy.turnYawMinDeg !== undefined) {
    out.turn = { minYaw: policy.turnYawMinDeg + 3, ...tuning.turn }
  }
  if (policy.neutralYawMaxDeg !== undefined) {
    // The centre gate should be at least as strict as the server's frontal
    // rule, and never looser than the local default.
    out.center = { maxYaw: Math.min(CHALLENGE_DEFAULTS.centerMaxYaw, policy.neutralYawMaxDeg), ...tuning.center }
  }
  return out
}

/**
 * Build the challenge objects for a server-issued list of names.
 *
 * `center` is always prepended — it is the framing gate and the source of the
 * neutral frame — and is stripped from `names` if the server sent it too.
 */
export function buildChallenges(
  names: ChallengeName[],
  tuning: ChallengeTuning = {},
): Challenge[] {
  const rest = names
    .filter((name) => name !== 'center')
    .map((name) => createChallenge(name, tuning))
  return [new CenterChallenge(tuning.center), ...rest]
}

function createChallenge(name: ChallengeName, tuning: ChallengeTuning): Challenge {
  switch (name) {
    case 'closeEyes':
      return new CloseEyesChallenge(tuning.closeEyes)
    case 'turnLeft':
      return new TurnLeftChallenge(tuning.turn)
    case 'turnRight':
      return new TurnRightChallenge(tuning.turn)
    case 'smile':
      return new SmileChallenge(tuning.smile)
    case 'openMouth':
      return new OpenMouthChallenge(tuning.openMouth)
    case 'nod':
      return new NodChallenge(tuning.nod)
    case 'center':
      return new CenterChallenge(tuning.center)
  }
}

File: sdk/android/liveness/src/main/java/com/ekyc/liveness/engine/Challenges.kt
package com.ekyc.liveness.engine

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.sign

enum class ChallengeName(val wire: String) {
    CENTER("center"),
    CLOSE_EYES("closeEyes"),
    TURN_LEFT("turnLeft"),
    TURN_RIGHT("turnRight"),
    SMILE("smile"),
    OPEN_MOUTH("openMouth"),
    NOD("nod"),
    MOVE_CLOSER("moveCloser"),
    MOVE_FARTHER("moveFarther");

    companion object {
        fun fromWire(s: String): ChallengeName = values().firstOrNull { it.wire == s }
            ?: throw IllegalArgumentException("unknown challenge: $s")
    }
}

enum class Direction { ABOVE, BELOW }

data class ChallengeMetric(val value: Float, val needed: Float, val direction: Direction)

/** Scratch memory for multi-phase challenges (a nod remembers which way it went). */
class ChallengeMemo {
    var nodSign: Int = 0
}

/**
 * Thresholds, and where they come from — a 1:1 port of `CHALLENGE_DEFAULTS`
 * in `@ekyc/react-native-ekyc` (see that file for the device measurements
 * behind every number). Every value is a *change from the person's own
 * neutral frame*.
 */
data class ChallengeTuning(
    val centerMaxYaw: Float = 12f,
    val centerMaxPitch: Float = 12f,
    /** Server rule 22° from neutral + 3° margin. Real turns measured 27–63°. */
    val turnMinYawDelta: Float = 25f,
    /** ML Kit: positive yaw = a turn to the user's own left (calibrated on device). */
    val yawSign: Int = -1,
    val nodMinPitchDelta: Float = 8f,
    val nodReturnFraction: Float = 0.4f,
    /** Mouth contour-gap rise over the neutral frame. Real opens measured 0.28–0.39 over ~0.04 shut. */
    val mouthOpenMinDelta: Float = 0.18f,
    val mouthOpenMinAbsolute: Float = 0.3f,
    val mouthCloseFraction: Float = 0.5f,
    val eyesClosedMaxOpen: Float = 0.5f,
    val eyesReopenMinOpen: Float = 0.7f,
    val smileMin: Float = 0.7f,
    val smileRelaxFraction: Float = 0.5f,
    val moveCloserMinGrow: Float = 0.25f,
    val moveFartherMinShrink: Float = 0.2f,
    val moveReturnBand: Float = 0.1f,
)

/**
 * One step of the flow. Predicates are judged against the neutral baseline;
 * multi-phase challenges need a movement *and its return* (eyes shut → open,
 * mouth open → closed, closer → back), so no single held pose or still photo
 * completes them.
 */
abstract class Challenge(protected val tuning: ChallengeTuning) {
    abstract val name: ChallengeName
    /** Hold required in the evidence phase; null = session default. 0 = event (one frame). */
    open val holdMs: Long? = null
    open val phaseCount: Int = 1
    /** Which phase the evidence frame is taken in (default: the last). */
    open val capturePhase: Int get() = phaseCount - 1
    /** Whether the head must come back to the centre before this step counts. */
    open val requiresRecenter: Boolean = true

    abstract fun isSatisfied(signal: FaceSignal, baseline: FaceSignal?, phase: Int, memo: ChallengeMemo): Boolean
    abstract fun metric(signal: FaceSignal, baseline: FaceSignal?, phase: Int, memo: ChallengeMemo): ChallengeMetric
}

class CenterChallenge(t: ChallengeTuning) : Challenge(t) {
    override val name = ChallengeName.CENTER
    override val requiresRecenter = false
    override fun isSatisfied(signal: FaceSignal, baseline: FaceSignal?, phase: Int, memo: ChallengeMemo) =
        abs(signal.yaw) <= tuning.centerMaxYaw && abs(signal.pitch) <= tuning.centerMaxPitch
    override fun metric(signal: FaceSignal, baseline: FaceSignal?, phase: Int, memo: ChallengeMemo) =
        ChallengeMetric(max(abs(signal.yaw), abs(signal.pitch)), tuning.centerMaxYaw, Direction.BELOW)
}

class CloseEyesChallenge(t: ChallengeTuning) : Challenge(t) {
    override val name = ChallengeName.CLOSE_EYES
    override val holdMs: Long = 0
    override val phaseCount = 2
    override val capturePhase = 0
    private fun open(s: FaceSignal) = (s.leftEye + s.rightEye) / 2f
    override fun isSatisfied(signal: FaceSignal, baseline: FaceSignal?, phase: Int, memo: ChallengeMemo) =
        if (phase == 0) open(signal) <= tuning.eyesClosedMaxOpen else open(signal) >= tuning.eyesReopenMinOpen
    override fun metric(signal: FaceSignal, baseline: FaceSignal?, phase: Int, memo: ChallengeMemo) =
        if (phase == 0) ChallengeMetric(open(signal), tuning.eyesClosedMaxOpen, Direction.BELOW)
        else ChallengeMetric(open(signal), tuning.eyesReopenMinOpen, Direction.ABOVE)
}

private fun turnDelta(signal: FaceSignal, baseline: FaceSignal?, sign: Int, left: Boolean): Float {
    val delta = signal.yaw - (baseline?.yaw ?: 0f)
    return if (left) -delta * sign else delta * sign
}

class TurnChallenge(t: ChallengeTuning, private val left: Boolean) : Challenge(t) {
    override val name = if (left) ChallengeName.TURN_LEFT else ChallengeName.TURN_RIGHT
    override fun isSatisfied(signal: FaceSignal, baseline: FaceSignal?, phase: Int, memo: ChallengeMemo) =
        turnDelta(signal, baseline, tuning.yawSign, left) >= tuning.turnMinYawDelta
    override fun metric(signal: FaceSignal, baseline: FaceSignal?, phase: Int, memo: ChallengeMemo) =
        ChallengeMetric(turnDelta(signal, baseline, tuning.yawSign, left), tuning.turnMinYawDelta, Direction.ABOVE)
}

class SmileChallenge(t: ChallengeTuning) : Challenge(t) {
    override val name = ChallengeName.SMILE
    override val phaseCount = 2
    override val capturePhase = 0
    override fun isSatisfied(signal: FaceSignal, baseline: FaceSignal?, phase: Int, memo: ChallengeMemo) =
        if (phase == 0) signal.smile >= tuning.smileMin else signal.smile <= tuning.smileMin * tuning.smileRelaxFraction
    override fun metric(signal: FaceSignal, baseline: FaceSignal?, phase: Int, memo: ChallengeMemo) =
        if (phase == 0) ChallengeMetric(signal.smile, tuning.smileMin, Direction.ABOVE)
        else ChallengeMetric(signal.smile, tuning.smileMin * tuning.smileRelaxFraction, Direction.BELOW)
}

/** The rigid-mask counter-measure: open (held) → closed again. */
class OpenMouthChallenge(t: ChallengeTuning) : Challenge(t) {
    override val name = ChallengeName.OPEN_MOUTH
    override val phaseCount = 2
    override val capturePhase = 0
    private fun needed(b: FaceSignal?) = if (b != null) b.mouthOpen + tuning.mouthOpenMinDelta else tuning.mouthOpenMinAbsolute
    private fun closedBelow(b: FaceSignal?) = (b?.mouthOpen ?: 0f) + tuning.mouthOpenMinDelta * tuning.mouthCloseFraction
    override fun isSatisfied(signal: FaceSignal, baseline: FaceSignal?, phase: Int, memo: ChallengeMemo) =
        if (phase == 0) signal.mouthOpen >= needed(baseline) else signal.mouthOpen <= closedBelow(baseline)
    override fun metric(signal: FaceSignal, baseline: FaceSignal?, phase: Int, memo: ChallengeMemo) =
        if (phase == 0) ChallengeMetric(signal.mouthOpen, needed(baseline), Direction.ABOVE)
        else ChallengeMetric(signal.mouthOpen, closedBelow(baseline), Direction.BELOW)
}

/** Nod: away from the resting pitch (either direction), then back near it. */
class NodChallenge(t: ChallengeTuning) : Challenge(t) {
    override val name = ChallengeName.NOD
    override val phaseCount = 2
    override val holdMs: Long = 0
    override fun isSatisfied(signal: FaceSignal, baseline: FaceSignal?, phase: Int, memo: ChallengeMemo): Boolean {
        val delta = signal.pitch - (baseline?.pitch ?: 0f)
        if (phase == 0) {
            if (abs(delta) < tuning.nodMinPitchDelta) return false
            memo.nodSign = sign(delta).toInt()
            return true
        }
        if (memo.nodSign == 0) return false
        return abs(delta) <= tuning.nodMinPitchDelta * tuning.nodReturnFraction
    }
    override fun metric(signal: FaceSignal, baseline: FaceSignal?, phase: Int, memo: ChallengeMemo): ChallengeMetric {
        val delta = abs(signal.pitch - (baseline?.pitch ?: 0f))
        return if (phase == 0) ChallengeMetric(delta, tuning.nodMinPitchDelta, Direction.ABOVE)
        else ChallengeMetric(delta, tuning.nodMinPitchDelta * tuning.nodReturnFraction, Direction.BELOW)
    }
}

private fun widthRatio(signal: FaceSignal, baseline: FaceSignal?): Float {
    val base = baseline?.box?.w ?: 0f
    return if (base > 1e-6f) signal.box.w / base else 1f
}

/** Move closer / farther: the face grows / shrinks by a fraction, then comes back. */
class MoveChallenge(t: ChallengeTuning, private val closer: Boolean) : Challenge(t) {
    override val name = if (closer) ChallengeName.MOVE_CLOSER else ChallengeName.MOVE_FARTHER
    override val phaseCount = 2
    override val capturePhase = 0
    override fun isSatisfied(signal: FaceSignal, baseline: FaceSignal?, phase: Int, memo: ChallengeMemo): Boolean {
        val r = widthRatio(signal, baseline)
        if (phase != 0) return abs(r - 1f) <= tuning.moveReturnBand
        return if (closer) r >= 1f + tuning.moveCloserMinGrow else r <= 1f - tuning.moveFartherMinShrink
    }
    override fun metric(signal: FaceSignal, baseline: FaceSignal?, phase: Int, memo: ChallengeMemo): ChallengeMetric {
        val r = widthRatio(signal, baseline)
        if (phase != 0) return ChallengeMetric(abs(r - 1f), tuning.moveReturnBand, Direction.BELOW)
        return if (closer) ChallengeMetric(r, 1f + tuning.moveCloserMinGrow, Direction.ABOVE)
        else ChallengeMetric(r, 1f - tuning.moveFartherMinShrink, Direction.BELOW)
    }
}

object Challenges {
    /** `center` is always first (framing gate + the neutral baseline), and stripped from [names] if present. */
    fun build(names: List<ChallengeName>, tuning: ChallengeTuning = ChallengeTuning()): List<Challenge> =
        listOf<Challenge>(CenterChallenge(tuning)) + names.filter { it != ChallengeName.CENTER }.map { create(it, tuning) }

    fun create(name: ChallengeName, tuning: ChallengeTuning): Challenge = when (name) {
        ChallengeName.CENTER -> CenterChallenge(tuning)
        ChallengeName.CLOSE_EYES -> CloseEyesChallenge(tuning)
        ChallengeName.TURN_LEFT -> TurnChallenge(tuning, left = true)
        ChallengeName.TURN_RIGHT -> TurnChallenge(tuning, left = false)
        ChallengeName.SMILE -> SmileChallenge(tuning)
        ChallengeName.OPEN_MOUTH -> OpenMouthChallenge(tuning)
        ChallengeName.NOD -> NodChallenge(tuning)
        ChallengeName.MOVE_CLOSER -> MoveChallenge(tuning, closer = true)
        ChallengeName.MOVE_FARTHER -> MoveChallenge(tuning, closer = false)
    }

    /** The pool the local flow draws from (every entry is a movement relative to the neutral frame). */
    val LOCAL_POOL = listOf(
        ChallengeName.CLOSE_EYES, ChallengeName.TURN_LEFT, ChallengeName.TURN_RIGHT,
        ChallengeName.OPEN_MOUTH, ChallengeName.MOVE_CLOSER, ChallengeName.MOVE_FARTHER,
    )

    /**
     * Same policy as the server's full tier: `openMouth` always (the challenge
     * a rigid mask cannot answer), the other slots random from the pool, the
     * whole list shuffled so a recording only replays by luck.
     */
    fun pickLocal(count: Int = 4, random: kotlin.random.Random = kotlin.random.Random.Default): List<ChallengeName> {
        val n = count.coerceIn(1, LOCAL_POOL.size)
        val rest = LOCAL_POOL.filter { it != ChallengeName.OPEN_MOUTH }.shuffled(random).take(n - 1)
        return (listOf(ChallengeName.OPEN_MOUTH) + rest).shuffled(random)
    }
}
